### ISSUE TITLE
jupyter macos: AttributeError: dlsym(RTLD_DEFAULT, GetMacOSStatusCommentString): symbol not found

### ISSUE THREAD

**Issue by idontgetoutmuch** [NixOS/nixpkgs#136485]
### Describe the bug
dlsym(RTLD_DEFAULT, GetMacOSStatusCommentString): symbol not found

### Steps To Reproduce
I have this problem at (which was HEAD a few minutes ago)

```
8c0e53465c2 *   origin/master Merge pull request #136474 from waldheinz/etc-file-source-to-store
```

```
[nix-shell:~/data-analytics]$ jupyter --version
jupyter --version
jupyter core     : 4.7.1
jupyter-notebook : 6.4.3
qtconsole        : 5.1.1
ipython          : 7.24.1
ipykernel        : 5.5.5
jupyter client   : 6.1.12
jupyter lab      : not installed
nbconvert        : 6.0.7
ipywidgets       : 7.6.3
nbformat         : 5.1.3
traitlets        : 5.0.5
```

```
[nix-shell:~/data-analytics]$ jupyter-notebook 
jupyter-notebook 
Traceback (most recent call last):
  File "/nix/store/9qwb5cqwj56ab19kyk205b98l2im4wgi-python3.9-notebook-6.4.3/bin/.jupyter-notebook-wrapped", line 6, in <module>
    from notebook.notebookapp import main
  File "/nix/store/9qwb5cqwj56ab19kyk205b98l2im4wgi-python3.9-notebook-6.4.3/lib/python3.9/site-packages/notebook/notebookapp.py", line 81, in <module>
    from .services.contents.filemanager import FileContentsManager
  File "/nix/store/9qwb5cqwj56ab19kyk205b98l2im4wgi-python3.9-notebook-6.4.3/lib/python3.9/site-packages/notebook/services/contents/filemanager.py", line 17, in <module>
    from send2trash import send2trash
  File "/nix/store/binrgnllvpf4pildj3wriczyvihrh92z-python3.9-Send2Trash-1.5.0/lib/python3.9/site-packages/send2trash/__init__.py", line 12, in <module>
    from .plat_osx import send2trash
  File "/nix/store/binrgnllvpf4pildj3wriczyvihrh92z-python3.9-Send2Trash-1.5.0/lib/python3.9/site-packages/send2trash/plat_osx.py", line 17, in <module>
    GetMacOSStatusCommentString = Foundation.GetMacOSStatusCommentString
  File "/nix/store/amn854jw94m1cph654w91pfyvz5r5a6h-python3-3.9.6/lib/python3.9/ctypes/__init__.py", line 387, in __getattr__
    func = self.__getitem__(name)
  File "/nix/store/amn854jw94m1cph654w91pfyvz5r5a6h-python3-3.9.6/lib/python3.9/ctypes/__init__.py", line 392, in __getitem__
    func = self._FuncPtr((name_or_ordinal, self))
AttributeError: dlsym(RTLD_DEFAULT, GetMacOSStatusCommentString): symbol not found
```

Using

```
with import <nixpkgs> {};

let
  py = pkgs.python3;
in

stdenv.mkDerivation rec {
  name = "python-env";

  buildInputs = let
    py = pkgs.python3;
  in [
    py
    py.pkgs.matplotlib
    py.pkgs.pandas
    py.pkgs.numpy
    py.pkgs.jupyter
  ];

  shellHook = ''
  echo "using python: ${py.name}"
  '';

}
```

and

```
[nix-shell:~/data-analytics]$ jupyter-notebook 
```

### Expected behavior
A jupyter notebook should appear.

### Additional context
```
System Software Overview:

  System Version:	macOS 11.5.1 (20G80)
  Kernel Version:	Darwin 20.6.0
```

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
dom@MacBook-Pro data-analytics % nix-shell -p nix-info --run "nix-info -m"
nix-shell -p nix-info --run "nix-info -m"
these paths will be fetched (0.01 MiB download, 0.03 MiB unpacked):
  /nix/store/b2b7irnx648iwz91smv0hk7gmhn9sipp-DarwinTools-1
  /nix/store/k83m68gphkk4iii7grz4n0h0j02qbrzg-nix-info
copying path '/nix/store/b2b7irnx648iwz91smv0hk7gmhn9sipp-DarwinTools-1' from 'https://cache.nixos.org'...
copying path '/nix/store/k83m68gphkk4iii7grz4n0h0j02qbrzg-nix-info' from 'https://cache.nixos.org'...
 - system: `"x86_64-darwin"`
 - host os: `Darwin 20.6.0, macOS 10.16`
 - multi-user?: `no`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.3.10`
 - channels(dom): `"nixpkgs-21.05pre278688.c0e88185200"`
 - nixpkgs: `/Users/dom/.nix-defexpr/channels/nixpkgs`
```

I don't know why it says macOS 10.16 - I am actually on 11.5.1

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

**Comment by idontgetoutmuch** [comment]
See also: https://github.com/jupyterlab/jupyterlab/issues/9863

**Comment by bergkvist** [comment]
nixpkgs uses a macOS Catalina build server to build Python, which is the reason for this problem (and probably why you see macOS 10.16.

I submitted a patch to Python which was just merged 4 days ago: https://github.com/python/cpython/pull/27251

It probably hasn't made it into nixpkgs yet. The change was backported to Python 3.8, 3.9, 3.10 and 3.11 - so it should also work with pkgs.python3 (Python 3.8 last time I checked).

**Comment by idontgetoutmuch** [comment]
Is there a way of knowing when it makes it into nixpkgs?

**Comment by bergkvist** [comment]
This issue will probably be closed when that happens: https://github.com/NixOS/nixpkgs/issues/105038

**Comment by jaapterwoerds** [comment]
I'm using an overlay to override the source and versions(3.8.12) for python38 and can confirm that works. Thanks for putting in the effort to get this fixed @bergkvist

**Comment by bergkvist** [comment]
There is an open PR which bumps the Python version here (https://github.com/NixOS/nixpkgs/pull/136254). When it gets merged this issue should get resolved. 

@jaapterwoerds Using an overlay with 3.8.11 or older would probably also work fine - as long as it doesn't hit the nixpkgs cache. The problem only manifests when you don't build and run Python on the same macOS system. In particular, when you build on macOS < 11 and run it on macOS >= 11. The nixpkgs cache is populated with Python binaries that have been built on macOS < 11.

**Comment by larsr** [comment]
TDLR; there is a quick-fix if you just want to get jupyter running.  Just do

    export DYLD_LIBRARY_PATH=/System/Library/Frameworks/Foundation.framework/Versions/C/Resources/BridgeSupport

I still have this problem on darwin after a recent `nixpkgs-unstable`.

## Recap

`jupyter notebook` fails because it imports `send2trash`, so the failure can quickly be verified with

    nix-shell --pure -p 'python39Packages.send2trash' --command 'python -m send2trash'

This prints out a trace that ends with

    AttributeError: dlsym(RTLD_DEFAULT, GetMacOSStatusCommentString): symbol not found

The fundamental problem is that [ctypes.util.find_library('Foundation') returns `None`, so LoadLibrary(find_library(...)) doesn't load anything](https://github.com/arsenetar/send2trash/blob/1.5.0/send2trash/plat_osx.py#L14).

By setting `DYLD_LIBRARY_PATH`, `find_library` will return the path, and the rest will work.

As a temporary fix, I set it in a `shellHook` in my `shell.nix` file with

```
with import <nixpkgs> {};
mkShell {
  buildInputs = [
    python39Packages.jupyter
 ];
  shellHook = ''
    export DYLD_LIBRARY_PATH=/System/Library/Frameworks/Foundation.framework/Versions/C/Resources/BridgeSupport
  '';
}
```

While this doesn't solve the problem longterm I hope this is useful for the people who come here in search of a quick solution.

**Comment by abathur** [comment]
> This issue will probably be closed when that happens: #105038

Closing this as complete along with the linked issue (see that thread for justification). Happy to reopen if I'm mistaken.
